### PR TITLE
BSDLib in Zephyr without TrustZone

### PR DESCRIPTION
**Pull request by sigvartmh**
I do know that it's a possibility we don't want this in nRF hence pointing my PR towards my own public fork. I just think it's cool to allow customers to use the chip to it's full extent without being weighed down by trustZone and having their RAM partitioned.

I'm not sure if the linker script generated is correct yet. Since I haven't hit the boundary between BSDLib and end of BSDLib but the general idea is too use a statically allocated block of memory which is placed in the `.bsdlib` section which is located in the `BSD_RESERVED_MEMORY_ADDRESS` and uses the `BSD_RESERVED_MEMORY_SIZE` to fit into the region.

Further more when initializing the BSD Library we calculate which regions needs to be marked non-secure through using those defines and the `NRF_SPU_S`. 

One advantage of doing this is that you'll be able to access CryptoCell functionality directly. Meaning you'll have an easier time using CryptoCell+mbedTLS+raw sockets to create a new TLS/DTLS session additionally to the one you have available in the modem.

### CODE REVIEW

**Comment by sigvartmh** [comment]
Tested and working sample of running zephyr without TrustZone with BSDLib

File: lib/bsdlib/bsdlib.c [hunk 1]
```diff
@@ -11,9 +11,79 @@
...
+	NRF_CLOCK_S->LFCLKSRC = CLOCK_LFCLKSRCCOPY_SRC_LFRC << CLOCK_LFCLKSRC_SRC_Pos;
+	NRF_CLOCK_S->EVENTS_LFCLKSTARTED  = 0;
+	NRF_CLOCK_S->TASKS_LFCLKSTART = 1;
+	while (NRF_CLOCK_S->EVENTS_LFCLKSTARTED == 0);
+	NRF_CLOCK_S->EVENTS_LFCLKSTARTED = 0;
+}
+
+static inline void setup_ns_power(void)
+{
+        NRF_SPU_S->PERIPHID[NRFX_PERIPHERAL_ID_GET(NRF_POWER_NS)].PERM = 0x80000102;
```
Comment (oyvindronningstad) [inline lib/bsdlib/bsdlib.c]: magic number

File: lib/bsdlib/bsdlib.c [hunk 2]
```diff
@@ -11,9 +11,79 @@
...
+        NRF_SPU_S->PERIPHID[NRFX_PERIPHERAL_ID_GET(NRF_POWER_S)].PERM = 0;
+}
+
+static inline void setup_s_ipc(void)
+{
+        NRF_SPU_S->PERIPHID[NRFX_PERIPHERAL_ID_GET(NRF_IPC_S)].PERM = 0x80000102;
```
Comment (oyvindronningstad) [inline lib/bsdlib/bsdlib.c]: magic number

File: lib/bsdlib/bsdlib.c [hunk 3]
```diff
@@ -23,7 +93,7 @@ static int init_ret;
 static int _bsdlib_init(struct device *unused)
 {
 	/* Setup the network IRQ used by the BSD library.
-	 * Note: No call to irq_enable() here, that is done through bsd_init().
+//	 * Note: No call to irq_enable() here, that is done through bsd_init().
```
Comment (oyvindronningstad) [inline lib/bsdlib/bsdlib.c]: Remove slashes

File: lib/bsdlib/bsdlib.ld [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#include <bsd_platform.h>
+__bsdlib = BSD_RESERVED_MEMORY_ADDRESS;
```
Comment (oyvindronningstad) [inline lib/bsdlib/bsdlib.ld]: I can't find where this is used.

File: samples/nrf9160/mqtt_simple_only_secure/src/main.c [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+#include <bsd.h>
+#include <lte_lc.h>
+#include <at_cmd.h>
+#include <at_notif.h>
+#include <net/mqtt.h>
+#include <net/socket.h>
+#include <net/bsdlib.h>
+#include <bsd_platform.h>
+
+static u8_t __attribute__((section (".bsdlib"), unused)) bsd_mem[BSD_RESERVED_MEMORY_SIZE];
```
Comment (oyvindronningstad) [inline samples/nrf9160/mqtt_simple_only_secure/src/main.c]: This is declared above as well. Is it meant to be declared twice?

File: lib/bsdlib/bsdlib.ld (hunk 4 above)
Comment (oyvindronningstad) [reply lib/bsdlib/bsdlib.ld]: If you want to place .bsdlib at BSD_RESERVED_MEMORY_ADDRESS, do

    . = BSD_RESERVED_MEMORY_ADDRESS;

Comment (sigvartmh) [reply lib/bsdlib/bsdlib.ld]: I see that placing it wasn't the easiest thing to do. I can't seem to get the location right with `zephyr_linker_sources` I want to place the static allocated memory block into the start of . = BSD_RESERVED_MEMORY_ADDRESS but I don't know if it is possible. I kind of want to put it before kernel_data_end if possible. I think we discussed this before and that the linker script language is not powerful enough to express put Y between X and Z.

File: samples/nrf9160/mqtt_simple_only_secure/src/main.c (hunk 5 above)
Comment (sigvartmh) [reply samples/nrf9160/mqtt_simple_only_secure/src/main.c]: No, forgot to remove this when I moved the other into the bsdlib source files.

File: lib/bsdlib/bsdlib.c (hunk 3 above)
Comment (sigvartmh) [reply lib/bsdlib/bsdlib.c]: Fixed.

File: lib/bsdlib/bsdlib.c (hunk 1 above)
Comment (sigvartmh) [reply lib/bsdlib/bsdlib.c]: Replaced

File: lib/bsdlib/bsdlib.c (hunk 2 above)
Comment (sigvartmh) [reply lib/bsdlib/bsdlib.c]: Replaced

File: lib/bsdlib/bsdlib.ld (hunk 4 above)
Comment (sigvartmh) [reply lib/bsdlib/bsdlib.ld]: The version 62c7147 at least creates the section at the correct address. I'm a bit unsure about the implication of having it that way. Also, it seems like the `static u8_t __attribute__((section (".bsdlib"), unused)) bsd_mem[BSD_RESERVED_MEMORY_SIZE];
` is not placed at all. Doing `u8_t __attribute__((section (".bsdlib"))) bsd_mem[BSD_RESERVED_MEMORY_SIZE];` worked fine but I hoped to not have it exposed.

**Comment by SebastianBoe** [comment]
I'm not familiar enough with Trustzone, cryptocell, or bsd lib to be able to say if we want this or not. But the code generally looks clean and in general we don't want to limit what users can do more than necessary so it sounds sane.

**Comment by sigvartmh** [comment]
@SebastianBoe Seems I currently have an issue with the placement of the unused/noinit section for the BSDLib memory section in the linker script. I really wished I was able to come up with a way of giving users access to the 64K of RAM before the BSD section and 256K - 49K - 64K after the BSD section.

Do you know a way of expressing this through the linker?

File: lib/bsdlib/Kconfig [hunk 6]
```diff
@@ -5,13 +5,17 @@
 
 menu "BSD Library for nrf91"
 
+config BSD_LIBRARY_ALLOW_SECURE
+	bool "Allow usage of BSD library in secure(Experimental)"
+	default n
```
Comment (hakonfam) [inline lib/bsdlib/Kconfig:10]: IMO we can skip this as it is the default

File: lib/bsdlib/bsdlib.c [hunk 7]
```diff
@@ -42,8 +121,12 @@ static int _bsdlib_init(struct device *unused)
 	return init_ret;
 }
 
+
 int bsdlib_init(void)
 {
+#ifdef CONFIG_BSD_LIBRARY_ALLOW_SECURE
+	setup_for_secure();
```
Comment (hakonfam) [inline lib/bsdlib/bsdlib.c:117]: Curious, why is this needed?

Comment (hakonfam) [reply lib/bsdlib/bsdlib.c:117]: If its more then 1 line just explain it IRL xD

File: samples/nrf9160/mqtt_simple_only_secure/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(hello_world)
+
+zephyr_linker_sources(SECTIONS app-data-section.ld)
+target_sources(app PRIVATE src/main.c)
```
Comment (hakonfam) [inline samples/nrf9160/mqtt_simple_only_secure/CMakeLists.txt]: Instead of copying the entire source dir, could you simply point to the source in the original sample folder, and then have the separate prj.conf in this folder as well. Assuming that there are no code chanes.

File: lib/bsdlib/bsdlib.c (hunk 7 above)
Comment (sigvartmh) [reply lib/bsdlib/bsdlib.c:117]: It's needed to setup the clock(BSDLib tries to enable the LF clock if it's not enabled) the modem is similar to nRF53 where it requires the RAM section used to be setup as non-secure(if not it won't be addressable on the BUS since the Modem is always a non-secure domain) Same with power it needs to be set as NS since BSDLib uses NS addressing same with the IPC. I think maybe we could allow POWER to be split. Sorry for the magic values seems they got reverted at some point.

So basically:
IPC NS
POWER NS or SPLIT
RAM NS

Are the requirements for operating with the modem from the secure domain.

We could also make
CLOCK NS and maybe it would work then we don't have to enable the LF clock and BSDLib could maybe handle it correctly if we allow Non-secure mapping. I'll have to test that.

File: samples/nrf9160/mqtt_simple_only_secure/CMakeLists.txt (hunk 8 above)
Comment (sigvartmh) [reply samples/nrf9160/mqtt_simple_only_secure/CMakeLists.txt]: The sample file is mostly for testing and just ensuring that it actually works I think I would remove it in favor of document how to build your application. So by using the `__in_section` we should be able to place whatever we prefer in the split RAM. Using https://docs.zephyrproject.org/latest/guides/code-relocation.html will actually allow you to place .bss/.text/.data sections in the split RAM also. So if you fill up your first 64K you should be able to place additional sections into `SRAM_AFTER_BSDLIB` 128K

File: lib/bsdlib/Kconfig (hunk 6 above)
Comment (sigvartmh) [reply lib/bsdlib/Kconfig:10]: Skip the whole option? We need something to notify BSDLib that you are now going to allow running it without TrustZone(the user knows what he's doing) option.

If not we could just remove the TrustZone requirement but I'm afraid we are creating a foot gun by allowing users who don't understand the implications of doing this that their RAM will be split etc.

Comment (hakonfam) [reply lib/bsdlib/Kconfig:10]: No not the option, just the `default n` part
